vllm/model_executor/models/llava.py

```diff
@@ -0,0 +1,245 @@
+from vllm.model_executor.layers.sampler import Sampler
+from vllm.model_executor.layers.vocab_parallel_embedding import ParallelLMHead
+from vllm.model_executor.models.llama import LlamaModel
+from vllm.model_executor.sampling_metadata import SamplingMetadata
+from vllm.model_executor.weight_utils import (default_weight_loader,
+                                              hf_model_weights_iterator)
+from vllm.sequence import SamplerOutput
+from vllm.config import VisionLanguageConfig, LoRAConfig
+
+KVCache = Tuple[torch.Tensor, torch.Tensor]
+
+if TYPE_CHECKING:
+    from transformers import LlavaConfig  # pylint: disable=ungrouped-imports
+    from vllm.model_executor.layers import LinearMethodBase  # pylint: disable=ungrouped-imports
+
+_KEYS_TO_MODIFY_MAPPING = {
+    "language_model.lm_head": "lm_head",
+    "language_model.model": "language_model",
+}
+
+
+# TODO(xwjiang): Run benchmark and decide if TP.
+class LlavaMultiModalProjector(nn.Module):
+
+    def __init__(self, vision_hidden_size: int, text_hidden_size: int,
+                 projector_hidden_act: str):
+        super().__init__()
+
+        self.linear_1 = nn.Linear(vision_hidden_size,
+                                  text_hidden_size,
+                                  bias=True)
+        self.act = get_act_fn(projector_hidden_act)
+        self.linear_2 = nn.Linear(text_hidden_size,
+                                  text_hidden_size,
+                                  bias=True)
+
+    def forward(self, image_features):
+        hidden_states = self.linear_1(image_features)
+        hidden_states = self.act(hidden_states)
+        hidden_states = self.linear_2(hidden_states)
+        return hidden_states
+
+
+def _merge_vision_embeddings(input_ids: torch.Tensor,
+                             inputs_embeds: torch.Tensor,
+                             vision_embeddings: torch.Tensor,
+                             image_token_id: int):
+    """In place merges in vision_embeddings with inputs_embeds."""
+    mask = (input_ids == image_token_id)
+    inputs_embeds[mask] = vision_embeddings.view(-1,
+                                                 vision_embeddings.shape[-1])
+
+
+# TODO(xwjiang): Add model correctness test. This model should produce the same
+# output as huggingface.
+class LlavaForConditionalGeneration(nn.Module):
+    supports_lora = False
+
+    def __init__(self,
+                 config: "LlavaConfig",
+                 vision_language_config: VisionLanguageConfig,
+                 linear_method: Optional["LinearMethodBase"] = None,
+                 lora_config: Optional["LoRAConfig"] = None) -> None:
+        # TODO(xwjiang): Remove when adding LoRA support.
+        assert not lora_config
+        super().__init__()
+        self.config = config
+
+        hf_text_config = config.text_config
+        hf_vision_config = config.vision_config
+        self.vision_language_config = vision_language_config
+
+        assert self.vision_language_config
```

Comment:
This can't fail if the type signature above is correct :)

Reply:
I would rather not rely on type hinting. I added some useful user facing information that will show up when someone does
```
llm = LLM(
    model="llava-hf/llava-1.5-7b-hf",
)
```
instead of 
```
llm = LLM(
    model="llava-hf/llava-1.5-7b-hf",
    image_input_type=VisionLanguageConfig.ImageInputType.PIXEL_VALUES,
    image_token_id=32000,
    image_input_shape=(1, 3, 336, 336),
    image_feature_size=576,
)
```